File: abyssal/skin/hidpi.py
"""Device scale for every derived-pixel cache.

On a scaled display (Omarchy at 1.6x) GTK hands the draw function a cairo
context whose target has a device scale. A cache surface allocated at LOGICAL
size is then upsampled on every blit - soft metal, soft type. Every cache in
this program therefore allocates through `surface()`, at device resolution,
and carries `scale()` in its key.

The host sets the scale once per frame from the real target. Headless QA
leaves it at 1.0 (or sets it explicitly to test HiDPI).
"""

from __future__ import annotations

import math

import cairo

_DS = 1.0


def set_scale(s: float) -> None:
    """Quantised to quarter steps so a fractional scale cannot churn caches."""
    global _DS
    _DS = max(1.0, min(4.0, round(float(s) * 4.0) / 4.0))


def scale() -> float:
    return _DS


def surface(w: float, h: float) -> cairo.ImageSurface:
    """An ARGB32 surface of LOGICAL size w x h at device resolution."""
    ds = _DS
    s = cairo.ImageSurface(cairo.FORMAT_ARGB32,
                           max(1, int(math.ceil(w * ds))),
                           max(1, int(math.ceil(h * ds))))
    s.set_device_scale(ds, ds)
    return s


def from_context(cr) -> float:
    """The device scale of a context's target (1.0 if it has none)."""
    try:
        return float(cr.get_target().get_device_scale()[0])
    except Exception:
        return 1.0

File: rust/src/skin/hidpi.rs
//! Device scale for every derived-pixel cache. Port of skin/hidpi.py.
//!
//! Every cache in this program allocates through `surface()`, at device
//! resolution, and carries `scale()` in its key. The host sets the scale once
//! per frame from the real target. Headless use leaves it at 1.0.

use std::cell::Cell;

use cairo::ImageSurface;

thread_local! {
    static DS: Cell<f64> = const { Cell::new(1.0) };
}

/// Quantised to quarter steps so a fractional scale cannot churn caches.
pub fn set_scale(s: f64) {
    let q = (s * 4.0).round() / 4.0;
    DS.with(|d| d.set(q.clamp(1.0, 4.0)));
}

pub fn scale() -> f64 {
    DS.with(|d| d.get())
}

/// An ARGB32 surface of LOGICAL size w x h at device resolution.
pub fn surface(w: f64, h: f64) -> ImageSurface {
    let ds = scale();
    let s = ImageSurface::create(
        cairo::Format::ARgb32,
        ((w * ds).ceil() as i32).max(1),
        ((h * ds).ceil() as i32).max(1),
    )
    .expect("hidpi surface");
    s.set_device_scale(ds, ds);
    s
}

/// The device scale of a context's target (1.0 if it has none).
pub fn from_context(cr: &cairo::Context) -> f64 {
    cr.target().device_scale().0
}
